enterprise/tailnet/coordinator.go

```diff
@@ -0,0 +1,507 @@
+
+	data, err := json.Marshal([]*agpl.Node{&node})
+	if err != nil {
+		return nil, xerrors.Errorf("marshal nodes: %w", err)
+	}
+
+	// Publish the new node to every listening socket.
+	var wg sync.WaitGroup
+	wg.Add(len(connectionSockets))
+	for _, connectionSocket := range connectionSockets {
+		connectionSocket := connectionSocket
+		go func() {
+			_ = connectionSocket.SetWriteDeadline(time.Now().Add(5 * time.Second))
+			_, _ = connectionSocket.Write(data)
+			wg.Done()
+		}()
+	}
+
+	wg.Wait()
+	return &node, nil
+}
+
+// Close closes all of the open connections in the coordinator and stops the
+// coordinator from accepting new connections.
+func (c *haCoordinator) Close() error {
+	c.mutex.Lock()
+	defer c.mutex.Unlock()
+
+	close(c.close)
+
+	wg := sync.WaitGroup{}
+
+	wg.Add(len(c.agentSockets))
+	for _, socket := range c.agentSockets {
+		socket := socket
+		go func() {
+			_ = socket.Close()
+			wg.Done()
+		}()
+	}
+
+	for _, connMap := range c.agentToConnectionSockets {
+		wg.Add(len(connMap))
+		for _, socket := range connMap {
+			socket := socket
+			go func() {
+				_ = socket.Close()
+				wg.Done()
+			}()
+		}
+	}
+
+	wg.Wait()
+	return nil
+}
+
+func (c *haCoordinator) publishNodesToAgent(recipient uuid.UUID, nodes []*agpl.Node) error {
+	msg, err := c.formatCallMeMaybe(recipient, nodes)
+	if err != nil {
+		return xerrors.Errorf("format publish message: %w", err)
+	}
+
+	err = c.pubsub.Publish("wireguard_peers", msg)
+	if err != nil {
+		return xerrors.Errorf("publish message: %w", err)
+	}
+
+	return nil
+}
+
+func (c *haCoordinator) publishAgentHello(id uuid.UUID) error {
+	msg, err := c.formatAgentHello(id)
+	if err != nil {
+		return xerrors.Errorf("format publish message: %w", err)
+	}
+
+	err = c.pubsub.Publish("wireguard_peers", msg)
+	if err != nil {
+		return xerrors.Errorf("publish message: %w", err)
+	}
+
+	return nil
+}
+
+func (c *haCoordinator) publishAgentToNodes(id uuid.UUID, node *agpl.Node) error {
+	msg, err := c.formatAgentUpdate(id, node)
+	if err != nil {
+		return xerrors.Errorf("format publish message: %w", err)
+	}
+
+	err = c.pubsub.Publish("wireguard_peers", msg)
+	if err != nil {
+		return xerrors.Errorf("publish message: %w", err)
+	}
+
+	return nil
+}
+
+func (c *haCoordinator) runPubsub() error {
+	cancelSub, err := c.pubsub.Subscribe("wireguard_peers", func(ctx context.Context, message []byte) {
+		sp := bytes.Split(message, []byte("|"))
+		if len(sp) != 4 {
+			c.log.Error(ctx, "invalid wireguard peer message", slog.F("msg", string(message)))
+			return
+		}
+
+		var (
+			coordinatorID = sp[0]
+			eventType     = sp[1]
+			agentID       = sp[2]
+			nodeJSON      = sp[3]
+		)
+
+		sender, err := uuid.ParseBytes(coordinatorID)
+		if err != nil {
+			c.log.Error(ctx, "invalid sender id", slog.F("id", string(coordinatorID)), slog.F("msg", string(message)))
+			return
+		}
+
+		// We sent this message!
+		if sender == c.id {
+			return
+		}
+
+		switch string(eventType) {
+		case "callmemaybe":
```

Comment:
i've seen a few of these do you think it makes sense to make this a constant or even a new string type?